sklearn/preprocessing/_csr_polynomial_expansion.pyx

```diff
@@ -1,73 +1,190 @@
-# Author: Andrew nystrom <awnystrom@gmail.com>
+# Authors: Andrew nystrom <awnystrom@gmail.com>
+#          Meekail Zain <zainmeekail@gmail.com>
+from libc.limits cimport INT_MAX, LONG_MAX
+from libc.math cimport sqrt, pow
+cimport numpy as cnp
+cnp.import_array()
 
 from scipy.sparse import csr_matrix
-cimport numpy as cnp
 import numpy as np
 
-cnp.import_array()
+ctypedef cnp.int8_t FLAG_t
 
-# TODO: use `cnp.{int,float}{32,64}` when cython#5230 is resolved:
-# https://github.com/cython/cython/issues/5230
-ctypedef fused DATA_T:
-    float
-    double
-    int
-    long
 
+# INDEX_{A,B}_t are defined to generate a proper Cartesian product
+# of types through Cython fused-type expansion.
+ctypedef fused INDEX_A_t:
+    cnp.int32_t
+    cnp.int64_t
+ctypedef fused INDEX_B_t:
+    cnp.int32_t
+    cnp.int64_t
 
-cdef inline cnp.int32_t _deg2_column(
-    cnp.int32_t d,
-    cnp.int32_t i,
-    cnp.int32_t j,
-    cnp.int32_t interaction_only,
+ctypedef fused DATA_t:
+    cnp.int32_t
+    cnp.int64_t
+    cnp.float32_t
+    cnp.float64_t
+
+cdef inline cnp.int64_t _deg2_column(
+    cnp.int64_t n_features,
+    cnp.int64_t i,
+    cnp.int64_t j,
+    FLAG_t interaction_only
 ) nogil:
     """Compute the index of the column for a degree 2 expansion
 
-    d is the dimensionality of the input data, i and j are the indices
+    n_features is the dimensionality of the input data, i and j are the indices
     for the columns involved in the expansion.
     """
+
+    # This is a conservative upper bound for the maximum value before the
+    # intermediate computation i**2 + 3*i overflows.
+    # Corresponds to to i**2 + 3*i = maxint32
+    cdef cnp.int64_t MAX_SAFE_INDEX_DEG2 = <cnp.int64_t> sqrt(LONG_MAX) - 4
```

Comment:
We can hardcode the value instead of computing it everytime:

```suggestion
    # Corresponds to to i**2 + 3*i = maxint32
    # i = sqrt(maxint32) - 4
    cdef cnp.int64_t MAX_SAFE_INDEX_DEG2 = 46340
```

This comments applies to the other bounds.